File: SEM1/TPJAD/materials/2. Java Core/legacy/1InterestingJavaExamples/JavaNet/ConvertBytes.java
import java.io.*;
public class ConvertBytes {
   
   private static final int lengthJava[] = {1, 2, 4, 8, 4, 8};
   
   private static int toByteArray(int t, Object source, byte[] dest, int start) {
      if ((source==null)||(dest==null)||(start<0)||(dest.length<start+lengthJava[t]))
         return 0;
      byte b[];
      ByteArrayOutputStream ob = new ByteArrayOutputStream();
      DataOutputStream o = new DataOutputStream(ob);
      try {
         switch (t) {
            case 0: o.writeByte(((Byte)source).byteValue()); break;
            case 1: o.writeShort(((Short)source).shortValue()); break;
            case 2: o.writeInt(((Integer)source).intValue()); break;
            case 3: o.writeLong(((Long)source).longValue()); break;
            case 4: o.writeFloat(((Float)source).floatValue()); break;
            case 5: o.writeDouble(((Double)source).doubleValue()); break;
         }//switch
      }//try
      catch (IOException e) {
         e.printStackTrace();
      }//catch
      b = ob.toByteArray();
      System.arraycopy(b, 0, dest, start, lengthJava[t]);
      return lengthJava[t];
   }//ConvertBytes.toByteArray

   private static int fromByteArray(int t, byte[] source, int start, Object[] dest) {
      if ((source==null)||(dest==null)||(start<0)||(source.length<start+lengthJava[t])||(dest.length<1))
         return 0;
      ByteArrayInputStream ib = new ByteArrayInputStream(source, start, lengthJava[t]);
      DataInputStream i = new DataInputStream(ib);
      Object ob = null;
      try {
         switch (t) {
            case 0: ob = new Byte(i.readByte()); break;
            case 1: ob = new Short(i.readShort()); break;
            case 2: ob = new Integer(i.readInt()); break;
            case 3: ob = new Long(i.readLong()); break;
            case 4: ob = new Float(i.readFloat()); break;
            case 5: ob = new Double(i.readDouble()); break;
         }//switch
      }//try
      catch (IOException e) {
         e.printStackTrace();
      }//catch
      dest[0] = ob;
      return lengthJava[t];
   }//ConvertBytes.fromByteArray
   
   public static int addByte(byte source, byte[] dest, int start) {
      return toByteArray(0, new Byte(source), dest, start);
   }//ConvertBytes.addByte
   public static int addShort(short source, byte[] dest, int start) {
      return toByteArray(1, new Short(source), dest, start);
   }//ConvertBytes.addShort
   public static int addInt(int source, byte[] dest, int start) {
      return toByteArray(2, new Integer(source), dest, start);
   }//ConvertBytes.addInt
   public static int addLong(long source, byte[] dest, int start) {
      return toByteArray(3, new Long(source), dest, start);
   }//ConvertBytes.addLong
   public static int addFloat(float source, byte[] dest, int start) {
      return toByteArray(4, new Float(source), dest, start);
   }//ConvertBytes.addFloat
   public static int addDouble(double source, byte[] dest, int start) {
      return toByteArray(5, new Double(source), dest, start);
   }//ConvertBytes.addFloat
   public static int getByte(byte[] source, int start, Byte[] dest) {
      return fromByteArray(0, source, start, dest);
   }//ConvertBytes.getByte
   public static int getShort(byte[] source, int start, Short[] dest) {
      return fromByteArray(1, source, start, dest);
   }//ConvertBytes.getShort
   public static int getInt(byte[] source, int start, Integer[] dest) {
      return fromByteArray(2, source, start, dest);
   }//ConvertBytes.getInt
   public static int getLong(byte[] source, int start, Long[] dest) {
      return fromByteArray(3, source, start, dest);
   }//ConvertBytes.getLong
   public static int getFloat(byte[] source, int start, Float[] dest) {
      return fromByteArray(4, source, start, dest);
   }//ConvertBytes.getFloat
   public static int getDouble(byte[] source, int start, Double[] dest) {
      return fromByteArray(5, source, start, dest);
   }//ConvertBytes.getDouble
   public static int getLength(int t) {
      if ((t<0)||(t>5))
         return 0;
      return lengthJava[t];
   }//ConvertBytes.getLength
}//ConvertBytes

File: SEM1/TPJAD/materials/2. Java Core/legacy/1InterestingJavaExamples/JavaNet/ConvertBytes.cpp
#include <stdio.h>
#include <string.h>
static int lengthJava[] = {1,2,4,8,4,8};
static int lengthC[] = {sizeof(char),sizeof(short),sizeof(int),
sizeof(long),sizeof(float),sizeof(double)};
static int first=1;

   int getLength(int t) {
   // Intoarce 0 la eroare de apel
   // >0 lungime reprezentare bigendian
   // <0 -lungime reprezentare littleendian
      if ((t<0)||(t>5))
         return 0;
      if (first) {
         int i;
          union {
            char c;
            int i;
         } x;
         first = 0;
         x.i = 1;
         if (x.c == 1)
            for (i=1; i<6; i++)
               lengthC[i] = -lengthC[i];
      }//if
      return lengthC[t];
   }//ConvertBytes.lengthOrd

   int C2Java(int t, char *source, char *dest) {
   // Intoarce 0 la eroare de apel
   // >0 lungimea transferata in dest
      int i, j, k;
      char c, copy[32];
      k = getLength(t);
      if ((k==0)||(source==NULL)||(dest==NULL))
         return 0;
      if (t==0) {
         dest[0] = source[0];
         return 1;
      }//if
      if (k<0) {
         k = -k;
         for (i=0, j=k-1; i<k; i++,j--)
            copy[i] = source[j];
      } 
      else
         memcpy(copy, source, k);
      if (k==lengthJava[t])
         memcpy(dest, copy, k);
      if (k>lengthJava[t]) {
         if (t>3)
            memcpy(dest, copy, lengthJava[t]);
         else
            memcpy(dest, copy+k-lengthJava[t], lengthJava[t]);
      }//if
      if (k<lengthJava[t]) {
         c = 0;
         if ((t<4)&&(copy[0]<0))
            c = 0xff;
         memset(dest, c, lengthJava[t]);
         if (t>3)
            memcpy(dest, copy, lengthJava[t]);
         else
            memcpy(dest+lengthJava[t]-k, copy, lengthJava[t]);
      }//if
      return lengthJava[t];
   }//ConvertBytes.C2Java

   int Java2C(int t, char *source, char *dest) {
   // Intoarce 0 la eroare de apel
   // >0 lungimea transferata in dest
      int i, j, k, x;
      char c, copy[32];
      k = getLength(t);
      if ((k==0)||(source==NULL)||(dest==NULL))
         return 0;
      if (t==0) {
         dest[0] = source[0];
         return 1;
      }//if
      x = 1;
      if (k<0) {
         x = -1;
         k = -k;
      }//if
      if (k==lengthJava[t])
         memcpy(copy, source, k);
      if (k<lengthJava[t]) {
         if (t>3)
            memcpy(copy, source, k);
         else
            memcpy(copy, source+lengthJava[t]-k, k);
      }//if
      if (k>lengthJava[t]) {
         c = 0;
         if ((t<4)&&(source[0]<0))
            c = 0xff;
         memset(copy, c, k);
         if (t>3)
            memcpy(copy, source, lengthJava[t]);
         else
            memcpy(copy, source+k-lengthJava[t], lengthJava[t]);
      }//if
      if (x>0)
         memcpy(dest, copy, k);
      else
         for (i=0, j=k-1; i<k; i++,j--)
            dest[i] = copy[j];
      return k;
   }//ConvertBytes.Java2C

    int addByte(char source, char *dest, int start) {
      return C2Java(0, (char *)&source, dest+start);
   }//ConvertBytes.addByte
    int addShort(short source, char *dest, int start) {
      return C2Java(1, (char *)&source, dest+start);
   }//ConvertBytes.addShort
    int addInt(int source, char *dest, int start) {
      return C2Java(2, (char *)&source, dest+start);
   }//ConvertBytes.addInt
    int addLong(long source, char *dest, int start) {
      return C2Java(3, (char *)&source, dest+start);
   }//ConvertBytes.addLong
    int addFloat(float source, char *dest, int start) {
      return C2Java(4, (char *)&source, dest+start);
   }//ConvertBytes.addFloat
    int addDouble(double source, char *dest, int start) {
      return C2Java(5, (char *)&source, dest+start);
   }//ConvertBytes.addDouble
    int getByte(char *source, int start, char *dest) {
      return Java2C(0, source+start, (char *)dest);
   }//ConvertBytes.getByte
    int getShort(char *source, int start, short *dest) {
      return Java2C(1, source+start, (char *)dest);
   }//ConvertBytes.getShort
    int getInt(char *source, int start, int *dest) {
      return Java2C(2, source+start, (char *)dest);
   }//ConvertBytes.getInt
    int getLong(char *source, int start, long *dest) {
      return Java2C(3, source+start, (char *)dest);
   }//ConvertBytes.getLong
    int getFloat(char *source, int start, float *dest) {
      return Java2C(4, source+start, (char *)dest);
   }//ConvertBytes.getFloat
    int getDouble(char *source, int start, double *dest) {
      return Java2C(5, source+start, (char *)dest);
   }//ConvertBytes.getDouble

